# metag_Pipeline
This pipeline evaluates the quality of genomic data before and after trimming and aligns the quality data to a target genome. This allows for the visualization communities within a sample, which is important to metagenomic analysis. Within the pipeline, several programs are used and therefore must be preloaded for the code to work. The programs used include: FastQC, MultiQC, Trimmomatic, Burrows-Wheeler Aligner (BWA), samtools, and Kraken2. 

# Included Tools and Packages
This tool comes in three parts:
  (1)metag_Pipeline.py
  (2)help_me.ppy
  (3)config_file
  (4)sample_config_file
 
 #Dependecies
 This pipeline requires Python, FASTQC, Trimmomatic, bwa, samtools, and  to be installed in order to run. 

The following is the conda environment recipe that includes these programs and what version this pipeline was developed in. 

| Firs|
| -------------------------- |
| Content Cell  | Content Cell  |
| Content Cell  | Content Cell  |

Note: channel ordering might be required to download these programs to a new or existing conda environment. 

Along with a configuration file, a help henu program should be included with this Pipeline. This help menu, called Help_Me.py, contains all help menus of all the programs used, recommended parameters, how to overcome common program problems, and additional sources of information. While the information has less detail than this document, the Help_Me file is an incredibly useful tool.

This information is taken directly from the help menu or the manual of each program. 


 
 
 
 
  #Usage 
  
  
  
  
  #
